kittinunf/Fuel
# Integrate Stetho

## Description
This PR adds Stetho integration to HttpURLConnection.

## How to test this?
1. Launch `sample/MainActivity`
2. Open `chrome://inspect/#devices` in your Chrome browser
3. Click `inspect`:

| |
| - |
|<img src="https://user-images.githubusercontent.com/1988156/48314080-ab6ce480-e607-11e8-8b0a-5df6ada8eaec.png" width="300px" /> |
4. Click `Go Couroutines` button.
- Note*: Due to the crash ([this issue](https://github.com/kittinunf/Fuel/issues/522)) in the sample app now, I cannot test different type of request other than `GET`.
5. Check the response:

![image](https://user-images.githubusercontent.com/1988156/48314108-030b5000-e608-11e8-8ae4-08b7bb0c70ff.png)

## How it works
According to [this issue](https://github.com/facebook/stetho/issues/461#issuecomment-264560920), 

For Stetho to work, we need to use [`StethoURLConnectionManager`](https://github.com/facebook/stetho/blob/master/stetho-urlconnection/src/main/java/com/facebook/stetho/urlconnection/StethoURLConnectionManager.java), with this dependency: `implementation 'com.facebook.stetho:stetho-urlconnection:1.5.0'`.

There are 4 methods that we need to call inside [HttpClient.kt](https://github.com/kittinunf/Fuel/blob/master/fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt):

- StethoManager#preConnect(connection: HttpURLConnection, requestEntity: SimpleRequestEntity?)
- StethoManager#interpretResponseStream(inputStream: InputStream)
- StethoManager#postConnect()
- StethoManager#httpExchangeFailed(ioException: IOException)

`StethoManager` is an Android dependency, so we cannot use this inside `HttpClient.kt` directly, so I expose the correct place for intercepting through `stethoHook`. Then I make another module to actually make the call for calling the 4 methods.

This way, only users that want to use `Stetho` needs to include this module.

## How to use it (user perspective)
To use `Stetho` for `Fuel`, they need to use the following steps:
1. add `compile 'com.github.worker8.fuel:fuel-stetho:<latest-version>'`
 - not sure if I should to `com.github.kittinunf.fuel` to make it consistent?
2. add it to `FuelManager`:
```
FuelManager.instance.apply {
            stethoHook = StethoHook("Fuel Sample App")
        }
```
3. Done! 🎉 

TODO:
- [x] use `stetho.httpExchangeFailed()` in the correct place
- [x] make `stethoHook` to become generic hook, so that it can be in a separate module





## Type of change

Check all that apply

- [ ] Refactoring (a change which changes the current internal or external interface)

## How Has This Been Tested?

In case you did not include tests describe why you and how you have verified the
changes, with instructions so we can reproduce. If you have added comprehensive
tests for your changes, you may omit this section.

## Checklist:

  - not hard to understand area
  - will do so if PR is ok 
  - since this is stetho integration, a good way to check it is to run `sample/MainActivity.kt`, then check `Stetho` output using `Chrome`
- [ ] Inspect the bytecode viewer, including reasoning why
  - how to do this? 😳 

Relevant issue: https://github.com/kittinunf/Fuel/issues/493
